File: irrexplorer/backends/lookingglass.py
"""
BGP Looking Glass backend integration.

Provides BGP routing information via looking glass servers.
Uses the lg.ring.nlnog.net API for route queries.
"""

import asyncio
import logging
from typing import Any, Dict, List, Optional

import aiohttp

logger = logging.getLogger(__name__)


class LookingGlassClient:
    """Client for NLNOG BGP Looking Glass queries."""

    def __init__(self, base_url: str = "https://lg.ring.nlnog.net"):
        """
        Initialize Looking Glass client.

        Args:
            base_url: Base URL for the looking glass (default: NLNOG Ring)
        """
        self.base_url = base_url.rstrip("/")
        self.timeout = aiohttp.ClientTimeout(total=30)

    async def query_prefix(self, prefix: str) -> Dict[str, Any]:
        """
        Query BGP routes for a specific prefix.

        Args:
            prefix: IP prefix to query (e.g., '192.0.2.0/24')

        Returns:
            Dictionary with routing information
        """
        try:
            async with aiohttp.ClientSession(timeout=self.timeout) as session:
                # NLNOG API uses query parameters
                url = f"{self.base_url}/api/prefix?q={prefix}&all=all"
                async with session.get(url) as response:
                    if response.status == 200:
                        data = await response.json()
                        return self._parse_prefix_response(data, prefix)
                    elif response.status == 404:
                        return {"prefix": prefix, "routes": [], "error": "Not found"}
                    else:
                        logger.error(f"Looking glass API error: {response.status}")
                        return {
                            "prefix": prefix,
                            "routes": [],
                            "error": f"API error: {response.status}",
                        }
        except asyncio.TimeoutError:
            logger.error(f"Timeout querying looking glass for prefix {prefix}")
            return {"prefix": prefix, "routes": [], "error": "Query timeout"}
        except Exception as e:
            logger.error(
                f"Error querying looking glass for prefix {prefix}: {e}",
                exc_info=True,
            )
            return {"prefix": prefix, "routes": [], "error": str(e)}

    async def query_asn(self, asn: int) -> Dict[str, Any]:
        """
        Query BGP routes for a specific ASN.

        Note: NLNOG LG doesn't have a direct ASN endpoint, so we return
        a message indicating limited support.

        Args:
            asn: Autonomous System Number

        Returns:
            Dictionary with ASN information
        """
        return {
            "asn": asn,
            "prefixes": [],
            "total_prefixes": 0,
            "as_name": None,
            "error": "ASN queries not supported by NLNOG Looking Glass. Use prefix queries instead.",
        }

    async def query_bgp_route(
        self, prefix: str, peer: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Query detailed BGP route information.

        Args:
            prefix: IP prefix to query
            peer: Optional specific peer to query

        Returns:
            Dictionary with detailed route information
        """
        try:
            async with aiohttp.ClientSession(timeout=self.timeout) as session:
                url = f"{self.base_url}/api/v1/route/{prefix}"
                if peer:
                    url += f"?peer={peer}"

                async with session.get(url) as response:
                    if response.status == 200:
                        data = await response.json()
                        return self._parse_route_response(data)
                    elif response.status == 404:
                        return {
                            "prefix": prefix,
                            "peer": peer,
                            "route": None,
                            "error": "Not found",
                        }
                    else:
                        logger.error(f"Looking glass API error: {response.status}")
                        return {
                            "prefix": prefix,
                            "peer": peer,
                            "route": None,
                            "error": f"API error: {response.status}",
                        }
        except asyncio.TimeoutError:
            logger.error(f"Timeout querying BGP route for {prefix}")
            return {
                "prefix": prefix,
                "peer": peer,
                "route": None,
                "error": "Query timeout",
            }
        except Exception as e:
            logger.error(f"Error querying BGP route for {prefix}: {e}", exc_info=True)
            return {
                "prefix": prefix,
                "peer": peer,
                "route": None,
                "error": str(e),
            }

    async def get_peers(self) -> List[Dict[str, Any]]:
        """
        Get list of available BGP peers.

        Returns:
            List of peer information dictionaries
        """
        try:
            async with aiohttp.ClientSession(timeout=self.timeout) as session:
                url = f"{self.base_url}/api/v1/peers"
                async with session.get(url) as response:
                    if response.status == 200:
                        data = await response.json()
                        return data.get("peers", [])
                    else:
                        logger.error(f"Looking glass API error: {response.status}")
                        return []
        except Exception as e:
            logger.error(f"Error fetching peers: {e}", exc_info=True)
            return []

    def _parse_prefix_response(
        self, data: Dict[str, Any], prefix: str
    ) -> Dict[str, Any]:
        """Parse prefix query response from NLNOG Looking Glass API."""
        routes = []

        # NLNOG API returns routes keyed by prefix
        routes_data = data.get("routes", {})
        for prefix_key, route_list in routes_data.items():
            for route in route_list:
                # Extract AS path from aspath field
                as_path = []
                if "aspath" in route and isinstance(route["aspath"], list):
                    as_path = [
                        asn[0] if isinstance(asn, list) else str(asn)
                        for asn in route["aspath"]
                    ]

                # Extract communities
                communities = []
                if "communities" in route and isinstance(route["communities"], list):
                    communities = [
                        comm[0] if isinstance(comm, list) else str(comm)
                        for comm in route["communities"]
                    ]

                routes.append(
                    {
                        "prefix": prefix_key,
                        "as_path": as_path,
                        "origin_asn": as_path[-1] if as_path else None,
                        "next_hop": route.get("exit_nexthop"),
                        "peer": route.get("ip"),
                        "communities": communities,
                        "local_pref": route.get("local_prf"),
                        "med": route.get("med"),
                    }
                )

        return {
            "prefix": prefix,
            "routes": routes,
            "total_routes": len(routes),
        }

    def _parse_asn_response(self, data: Dict[str, Any], asn: int) -> Dict[str, Any]:
        """Parse ASN query response."""
        prefixes = []
        for prefix_data in data.get("prefixes", []):
            prefixes.append(
                {
                    "prefix": prefix_data.get("prefix"),
                    "origin": prefix_data.get("origin"),
                    "peers": prefix_data.get("peers", []),
                }
            )

        return {
            "asn": asn,
            "prefixes": prefixes,
            "total_prefixes": len(prefixes),
            "as_name": data.get("as_name"),
        }

    def _parse_route_response(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Parse detailed route query response."""
        route = data.get("route")
        if not route:
            return {"route": None}

        return {
            "prefix": route.get("prefix"),
            "as_path": route.get("as_path", []),
            "origin_asn": (
                route.get("as_path", [])[-1] if route.get("as_path") else None
            ),
            "next_hop": route.get("next_hop"),
            "peer": route.get("peer"),
            "communities": route.get("communities", []),
            "local_pref": route.get("local_pref"),
            "med": route.get("med"),
            "atomic_aggregate": route.get("atomic_aggregate", False),
            "aggregator": route.get("aggregator"),
            "originator_id": route.get("originator_id"),
            "cluster_list": route.get("cluster_list", []),
        }


# Global client instance
_looking_glass_client: Optional[LookingGlassClient] = None


def get_looking_glass_client() -> LookingGlassClient:
    """Get or create the global looking glass client instance."""
    global _looking_glass_client
    if _looking_glass_client is None:
        _looking_glass_client = LookingGlassClient()
    return _looking_glass_client

File: go-backend/internal/datasources/lookingglass.go
package datasources

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

type LookingGlassClient struct {
	httpClient *http.Client
	baseURL    string
}

func NewLookingGlassClient(baseURL string, timeout time.Duration) *LookingGlassClient {
	if baseURL == "" {
		baseURL = "https://lg.ring.nlnog.net"
	}
	return &LookingGlassClient{
		httpClient: &http.Client{Timeout: timeout},
		baseURL:    trimTrailingSlash(baseURL),
	}
}

func (c *LookingGlassClient) QueryPrefix(ctx context.Context, prefix string) map[string]any {
	data, err := c.get(ctx, fmt.Sprintf("%s/api/prefix?q=%s&all=all", c.baseURL, url.QueryEscape(prefix)))
	if err != nil {
		return map[string]any{"prefix": prefix, "routes": []any{}, "error": err.Error()}
	}
	routes := make([]map[string]any, 0)
	if routesData, ok := data["routes"].(map[string]any); ok {
		for prefixKey, rawList := range routesData {
			for _, rawRoute := range toAnySlice(rawList) {
				route, ok := rawRoute.(map[string]any)
				if !ok {
					continue
				}
				asPath := flattenPath(route["aspath"])
				routes = append(routes, map[string]any{
					"prefix":      prefixKey,
					"as_path":     asPath,
					"origin_asn":  lastOrNil(asPath),
					"next_hop":    route["exit_nexthop"],
					"peer":        route["ip"],
					"communities": flattenPath(route["communities"]),
					"local_pref":  route["local_prf"],
					"med":         route["med"],
				})
			}
		}
	}
	return map[string]any{"prefix": prefix, "routes": routes, "total_routes": len(routes)}
}

func (c *LookingGlassClient) QueryASN(ctx context.Context, asn int) map[string]any {
	data, err := c.get(ctx, fmt.Sprintf("https://stat.ripe.net/data/announced-prefixes/data.json?resource=AS%d", asn))
	if err != nil {
		return map[string]any{"asn": asn, "prefixes": []any{}, "error": err.Error()}
	}
	prefixes := make([]map[string]any, 0)
	dataObj, _ := data["data"].(map[string]any)
	for _, raw := range toAnySlice(dataObj["prefixes"]) {
		item, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		prefixes = append(prefixes, map[string]any{
			"prefix": item["prefix"],
			"origin": fmt.Sprintf("AS%d", asn),
			"peers":  []any{},
		})
	}
	return map[string]any{"asn": asn, "prefixes": prefixes, "total_prefixes": len(prefixes), "as_name": nil}
}

func (c *LookingGlassClient) QueryRoute(ctx context.Context, prefix, peer string) map[string]any {
	endpoint := fmt.Sprintf("%s/api/v1/route/%s", c.baseURL, url.PathEscape(prefix))
	if peer != "" {
		endpoint += "?peer=" + url.QueryEscape(peer)
	}
	data, err := c.get(ctx, endpoint)
	if err != nil {
		return map[string]any{"prefix": prefix, "peer": peer, "route": nil, "error": err.Error()}
	}
	route, _ := data["route"].(map[string]any)
	if route == nil {
		return map[string]any{"route": nil}
	}
	path := flattenPath(route["as_path"])
	return map[string]any{
		"prefix":           route["prefix"],
		"as_path":          path,
		"origin_asn":       lastOrNil(path),
		"next_hop":         route["next_hop"],
		"peer":             route["peer"],
		"communities":      toAnySlice(route["communities"]),
		"local_pref":       route["local_pref"],
		"med":              route["med"],
		"atomic_aggregate": route["atomic_aggregate"],
		"aggregator":       route["aggregator"],
		"originator_id":    route["originator_id"],
		"cluster_list":     toAnySlice(route["cluster_list"]),
	}
}

func (c *LookingGlassClient) Peers(ctx context.Context) []any {
	data, err := c.get(ctx, fmt.Sprintf("%s/api/v1/peers", c.baseURL))
	if err != nil {
		return []any{}
	}
	return toAnySlice(data["peers"])
}

func (c *LookingGlassClient) get(ctx context.Context, endpoint string) (map[string]any, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return nil, err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API error: %d", resp.StatusCode)
	}
	var data map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}
	return data, nil
}

func flattenPath(value any) []any {
	items := make([]any, 0)
	for _, raw := range toAnySlice(value) {
		if nested, ok := raw.([]any); ok && len(nested) > 0 {
			items = append(items, nested[0])
		} else {
			items = append(items, raw)
		}
	}
	return items
}

func lastOrNil(items []any) any {
	if len(items) == 0 {
		return nil
	}
	return items[len(items)-1]
}

func trimTrailingSlash(value string) string {
	for len(value) > 0 && value[len(value)-1] == '/' {
		value = value[:len(value)-1]
	}
	return value
}
